src/org/rust/lang/core/psi/util/RustModExtensions.kt

```diff
@@ -0,0 +1,61 @@
+package org.rust.lang.core.psi.util
+
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RustCompositeElement
+import org.rust.lang.core.psi.RustCrateItem
+import org.rust.lang.core.psi.RustModItem
+
+
+private val RustModItem.modDir: PsiDirectory?
+    get() {
+        val parent = parentModInFile ?: return containingFile.parent
+
+        // FIXME(matklad): mod name should be not null, should we rearange our rules for file/mod?
+        val name = this.name ?: return null
+
+        return parent.modDir?.findSubdirectory(name)
+    }
+
+
+val RustCompositeElement.parentModInFile: RustModItem?
+    get() = parentOfType<RustModItem>()
+
+private val RustModItem.isCrateRoot: Boolean
+    get() = parentModInFile == null &&
+        (containingFile.name == "main.rs" || containingFile.name == "lib.rs")
+
+private val RustModItem.ownsDirectory: Boolean
+    get() = parentModInFile != null || // any inline nested module owns a directory
+        containingFile.name == "mod.rs" ||
```

Comment:
I'm not 100% sure that I use correct term, but I call `mod foo {  }` and inline module and `mod foo;` a module declaration.

Hm, if you have 

```Rust
mod foo {
  mod bar { /* ... */ }
}
```

inside a file called `foo.rs` then you have two different nested modules with name `foo`. That is canonical path for `bar` will end in `foo::foo::bar`.  I think you would rather mean

```Rust
//foo.rs
mod bar {

}
```

Either way, my answer stays the same: `mod bar { }` is a directory owner, because it has `{ }`.